compiler/src/iree/compiler/Dialect/Util/Transforms/Patterns.cpp

```diff
@@ -496,6 +571,308 @@ struct MergeIndexSwitchPattern : public OpRewritePattern<scf::IndexSwitchOp> {
+                        rewriter, ifOp.getLoc(), ifOp.getResultTypes()));
+        } else {
+          rewriter.eraseOp(ifOp);
+        }
+      }
+      return success();
+    }
+  }
+};
+
+// Simplifies scf.while when the body contains unreachable.
+struct SimplifyWhileWithUnreachablePattern
+    : public OpRewritePattern<scf::WhileOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(scf::WhileOp whileOp,
+                                PatternRewriter &rewriter) const override {
+    // Check if the after region (loop body) is unreachable.
+    if (whileOp.getAfter().empty()) {
+      return failure();
+    }
+    Block *afterBlock = whileOp.getAfterBody();
+    if (!isBlockUnreachable(afterBlock)) {
+      return failure();
+    }
+
+    // If the loop body is unreachable it means the loop condition must
+    // always be false on the first iteration (otherwise we'd hit unreachable).
+    // We can replace the entire loop with the values that would be returned
+    // from the condition region when the condition is false.
+    // For scf.while the results come from the scf.condition operands and
+    // not from the inits directly.
+    if (whileOp.getNumResults() > 0) {
+      // The condition must evaluate to false on the first iteration so we
+      // simulate executing just the before region with the init values.
+      // We need to map the block arguments to the init values and then
+      // get what the condition would yield.
+      SmallVector<Value> resultValues;
+      Block *beforeBlock = whileOp.getBeforeBody();
+      auto condOp = cast<scf::ConditionOp>(beforeBlock->getTerminator());
+      IRMapping mapping;
+      for (auto [blockArg, initVal] :
+           llvm::zip(beforeBlock->getArguments(), whileOp.getInits())) {
+        mapping.map(blockArg, initVal);
+      }
+      for (Value condOperand : condOp.getArgs()) {
+        resultValues.push_back(mapping.lookupOrDefault(condOperand));
+      }
+      rewriter.replaceOp(whileOp, resultValues);
+    } else {
+      rewriter.eraseOp(whileOp);
+    }
+    return success();
+  }
+};
+
+// Simplifies scf.index_switch when cases contain unreachable.
+struct SimplifyIndexSwitchWithUnreachablePattern
+    : public OpRewritePattern<scf::IndexSwitchOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(scf::IndexSwitchOp switchOp,
+                                PatternRewriter &rewriter) const override {
+    // Collect which cases are unreachable.
+    SmallVector<int64_t> remainingCases;
+    SmallVector<Region *> remainingRegions;
+    bool hasChanges = false;
+    for (auto [idx, caseRegion] : llvm::enumerate(switchOp.getCaseRegions())) {
+      if (!caseRegion.empty()) {
+        Block &block = caseRegion.front();
+        // Check if this case contains unreachable.
+        if (isBlockUnreachable(&block)) {
+          hasChanges = true;
+          continue; // skip this case
```

Comment:
optional nit: I'd combine the comments. E.g., `// Skip this case if it contains an unreachable op.`